src/cryptonote_core/service_node_quorum_cop.cpp

```diff
@@ -420,9 +445,34 @@ namespace service_nodes
     if (m_uptime_proof_seen[pubkey].timestamp >= now - (UPTIME_PROOF_FREQUENCY_IN_SECONDS / 2))
       return false; // already received one uptime proof for this node recently.
 
-    crypto::hash hash = make_hash(pubkey, timestamp);
-    if (!crypto::check_signature(hash, pubkey, sig))
+    crypto::hash hash;
+
+    const uint64_t hf12_height = m_core.get_earliest_ideal_height_for_version(cryptonote::network_version_12_checkpointing);
+
+    const bool enforce_v2 = (hf12_height != std::numeric_limits<uint64_t>::max() && height >= hf12_height + 2);
+
+    /// Accept both old and new uptime proofs in a small window of 2 blocks
+    /// after switching to hf 12; (for simplicity accept new signatures before hf 12 too)
+    bool signature_ok = false;
+    if (!enforce_v2) {
+
+      hash = make_hash(pubkey, timestamp);
+
+      signature_ok = crypto::check_signature(hash, pubkey, sig);
+
+      if (!signature_ok) {
+        hash = make_hash_v2(pubkey, timestamp, proof.storage_endpoint);
+        signature_ok = crypto::check_signature(hash, pubkey, sig);
+      }
+
+    } else {
+      hash = make_hash_v2(pubkey, timestamp, proof.storage_endpoint);
+      signature_ok = crypto::check_signature(hash, pubkey, sig);
+    }
+
```

Comment:
No validation on the ipv4 network address passed? Actually could be a problem if someone passed around localhost and then clients programatically query localhost for the service node.

Reply:
The ip is validated on daemon startup, so not strictly necessary to validate here?

Reply:
Like a malicious service node purposely puts in loopback in the proof and then signs and submits it

Reply:
OK, added sanity check.